### PR TITLE
Add FXIOS-10617 [ToS] Add Glean usage profile id

### PR DESCRIPTION
**Pull request by thatswinnie**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10617)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/23407)

## :bulb: Description
Adds the Glean usage profile id.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/metrics.yaml [hunk 1]
```diff
@@ -709,6 +709,27 @@ app:
...
+    data_reviews:
+      - https://github.com/mozilla-mobile/firefox-ios/pull/
+    data_sensitivity:
+      - technical
+      - highly_sensitive
+    notification_emails:
+      - fx-ios-data-stewards@mozilla.com
+    expires: never
+    send_in_pings:
+      - dau-reporting
```
Comment (badboy) [inline firefox-ios/Client/metrics.yaml:731]:
```suggestion
      - usage-reporting
```

File: firefox-ios/Client/metrics.yaml [hunk 2]
```diff
@@ -709,6 +709,27 @@ app:
...
+      not shared with other telemetry data.
+    bugs:
+      - https://github.com/mozilla-mobile/firefox-ios/pull/
+    data_reviews:
+      - https://github.com/mozilla-mobile/firefox-ios/pull/
+    data_sensitivity:
+      - technical
+      - highly_sensitive
+    notification_emails:
+      - fx-ios-data-stewards@mozilla.com
```
Comment (badboy) [inline firefox-ios/Client/metrics.yaml:728]: Can you add `glean-team@mozilla.com` here too?
```suggestion
      - fx-ios-data-stewards@mozilla.com
      - glean-team@mozilla.com
```

File: firefox-ios/Client/Frontend/Settings/Main/Support/SendAnonymousUsageDataSetting.swift [hunk 3]
```diff
@@ -54,6 +55,11 @@ class SendAnonymousUsageDataSetting: BoolSetting {
         self.settingDidChange = { [weak self] value in
             // AdjustHelper.setEnabled($0)
             DefaultGleanWrapper.shared.setUpload(isEnabled: value)
+
+            if !value {
+                self?.prefs?.removeObjectForKey(PrefsKeys.Usage.profileId)
```
Comment (badboy) [inline firefox-ios/Client/Frontend/Settings/Main/Support/SendAnonymousUsageDataSetting.swift:60]: This should also reset the profile id, just to ensure we _really_ forget the previous one.

```suggestion
                self?.prefs?.removeObjectForKey(PrefsKeys.Usage.profileId)
                GleanMetrics.Usage.profileId.set(UUID(uuidString: "beefbeef-beef-beef-beef-beeefbeefbee"))
```

File: firefox-ios/Client/Telemetry/TelemetryWrapper.swift [hunk 4]
```diff
@@ -104,6 +104,14 @@ class TelemetryWrapper: TelemetryWrapperProtocol, FeatureFlaggable {
             GleanMetrics.LegacyIds.clientId.set(uuid)
         }
 
+        // Set or generate profile id used for usage reporting
+        if let uuidString = profile.prefs.stringForKey(PrefsKeys.Usage.profileId), let uuid = UUID(uuidString: uuidString) {
+            GleanMetrics.Usage.profileId.set(uuid)
+        } else {
+            let uuid = GleanMetrics.Usage.profileId.generateAndSet()
+            profile.prefs.setString(uuid.uuidString, forKey: PrefsKeys.Usage.profileId)
+        }
```
Comment (badboy) [inline firefox-ios/Client/Telemetry/TelemetryWrapper.swift:113]: This should only happen if `sendUsageData == true`.
If `false` we should also set the canary value `"beefbeef-beef-beef-beef-beeefbeefbee"` (as above)

**Comment by thatswinnie** [comment]
@adudenamedruby 

# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?
_It will give us the ability to count profiles._

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?
_Understand usage of Firefox iOS. This change allows us to offer Firefox users more granular choices in data collection._

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?
_There are no other methods_

4) Can current instrumentation answer these questions?
_No._

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>usage.profile_id<br>Records a UUID uniquely identifying the profile</td>
    <td>1 + 4</td>
    <td>FXIOS-10617</td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
[Firefox iOS Telemetry Dictionary](https://dictionary.telemetry.mozilla.org/apps/firefox_ios)

7) How long will this data be collected?  Choose one of the following:
_Set to never expire_

8) What populations will you measure?
* Which release channels? _All_
* Which countries? _All_
* Which locales? _All_
* Any other filters?  Please describe in detail below.

9) If this data collection is default on, what is the opt-out mechanism for users?
_The user can opt out_

10) Please provide a general description of how you will analyze this data.
_The data will be analyzed using Glean._

11) Where do you intend to share the results of your analysis?
_The results will be shared internally through Glean and with advertising partners._

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? If so:
_No third-party tools are proposed for this data collection._

**Review by travis79 — APPROVED** [review]
> @adudenamedruby @travis79
> # Request for data collection review form
> 
> **All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**
> 
>     1. What questions will you answer with this data?
>        _It will give us the ability to count profiles._
> 
>     2. Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?
>        _Understand usage of Firefox iOS. This change allows us to offer Firefox users more granular choices in data collection._
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
>        _There are no other methods_
> 
>     4. Can current instrumentation answer these questions?
>        _No._
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.
> 
> 
> **Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**
> Measurement Description 	Data Collection Category 	Tracking Bug #
> usage.profile_id
> Records a UUID uniquely identifying the profile 	1 + 4 	[FXIOS-10617](https://mozilla-hub.atlassian.net/browse/FXIOS-10617)
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
>        [Firefox iOS Telemetry Dictionary](https://dictionary.telemetry.mozilla.org/apps/firefox_ios)
> 
>     7. How long will this data be collected?  Choose one of the following:
>        _Set to never expire_
> 
>     8. What populations will you measure?
> 
> 
>     * Which release channels? _All_
> 
>     * Which countries? _All_
> 
>     * Which locales? _All_
> 
>     * Any other filters?  Please describe in detail below.
> 
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
>        _The user can opt out_
> 
>     10. Please provide a general description of how you will analyze this data.
>         _The data will be analyzed using Glean._
> 
>     11. Where do you intend to share the results of your analysis?
>         _The results will be shared internally through Glean and with advertising partners._
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? If so:
>         _No third-party tools are proposed for this data collection._

This has been approved by Mozilla Trust & Legal, I'm providing the data-review+ here as covered by their review and approval.

File: firefox-ios/Client/Telemetry/TelemetryWrapper.swift [hunk 5]
```diff
@@ -100,10 +100,25 @@ class TelemetryWrapper: TelemetryWrapperProtocol, FeatureFlaggable {
...
 
         // Get the legacy telemetry ID and record it in Glean for the deletion-request ping
-        if let uuidString = UserDefaults.standard.string(forKey: "telemetry-key-prefix-clientId"), let uuid = UUID(uuidString: uuidString) {
+        if let uuidString = UserDefaults.standard.string(forKey: "telemetry-key-prefix-clientId"),
+           let uuid = UUID(uuidString: uuidString) {
             GleanMetrics.LegacyIds.clientId.set(uuid)
         }
 
+        // Set or generate profile id used for usage reporting
+        if profile.prefs.boolForKey(AppConstants.prefSendUsageData) ?? true {
```
Comment (badboy) [inline firefox-ios/Client/Telemetry/TelemetryWrapper.swift:109]: this is passed in as `sendUsageData`:

```suggestion
        if sendUsageData {
```

**Review by badboy — APPROVED** [review]
one last nit, but then good to go!

**Review by MattLichtenstein — APPROVED** [review]

File: firefox-ios/Client/Frontend/Settings/Main/Support/SendDataSetting.swift [hunk 6]
```diff
@@ -80,6 +81,16 @@ class SendDataSetting: BoolSetting {
         self.settingDidChange = { [weak self] value in
             // AdjustHelper.setEnabled($0)
             DefaultGleanWrapper.shared.setUpload(isEnabled: value)
+
+            if !value {
+                self?.prefs?.removeObjectForKey(PrefsKeys.Usage.profileId)
+
+                // set dummy uuid to make sure the previous one is deleted
+                if let uuid = UUID(uuidString: "beefbeef-beef-beef-beef-beeefbeefbee") {
```
[the remark this answers is not in the document]
Comment (lmarceau) [reply firefox-ios/Client/Frontend/Settings/Main/Support/SendDataSetting.swift:89]: I didn't knew Winnie liked beef that much

[the remark this answers is not in the document]
Comment (badboy) [reply firefox-ios/Client/Frontend/Settings/Main/Support/SendDataSetting.swift:89]: c0ffee (and decaf) were already taken and I couldn't come up with other fun stuff.

[the remark this answers is not in the document]
Comment (thatswinnie) [reply firefox-ios/Client/Frontend/Settings/Main/Support/SendDataSetting.swift:89]: LOL. I like beef and I cannot lie 😇 
But seriously, that string actually came from @badboy

**Pull request merged**